proposed/2023/Add-dotnet-search.md

```diff
@@ -0,0 +1,91 @@
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+Imagine you have your dotnet CLI open. You would like to look up a NuGet package named `MyPackage` from a source `<MySource>`. No worries, you can just use the following command : `dotnet package search MyPackage --source <MySource>`. It will provide you with the list of the packages in the source `<MySource>` that match with the search criteria.
+
+The `package search [search terms] [options]` command will have the following options
+
+| Option | Function |
+|---------|:----------|
+| `--source` | A source to search |
+| `--exact-match` | Return exact matches only as a search result |
+| `--prerelease` | Allow prerelease packages to be shown. |
+| `--interactive` | Allows the command to stop and wait for user input or action (for example to complete authentication).|
+| `--take` | The number of results to return. The default value is 20.|
+| `--help` | Show command help and usage information |
+|||
+
+#### **Option `--source`**
+
+This option will specify a list of sources to search from. If a source is not specified using this option, the sources in the `nuget.config` file will be used.
+
+#### **Option `-exact-match`**
+
+- This option will allow for users to be able to search and have only exact matches as an output.
+- For example if a user uses `dotnet package search NuGet.CommandLine`
+
+        >NuGet.CommandLine | 6.7.0 | Downloads: N/A
+        >NuGet.CommandLine.XPlat | 6.7.0 | Downloads: N/A
+        >NuGet.Commands | 6.7.0 | Downloads: N/A
+        >NuGet.exe | 3.4.3 | Downloads: N/A
+        >NuGet.Bootstrapper | 2.6.0 | Downloads: N/A
+        >CommandLineParser20 | 2.0.0 | Downloads: N/A
+        >NuGet.VerifyMicrosoftPackage | 1.0.0 | Downloads: N/A
+        >NuGet.for.MSBuild | 2.1.0 ...
+
+- Using ``dotnet package search NuGet.CommandLine --exact-match`` on the other side will have the following output
+
+         >NuGet.CommandLine | 6.7.0 | Downloads: N/A
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+This command will prepare a search parameter using the inputs of  `Search Term`, `--take`, and `--prerelease`. This parameter is then used prepare an API query to the specified source. In the dotnet/sdk repo, there is already a command, `dotnet tool search`, which does a similar thing. However, it specifies the package type to be only `dotnettool`. Instead of adding a new set of classes and methods, I will modify the methods and classes used by `dotnet tool search` to do a general search when needed. This will allow both `dotnet package search` and `dotnet tool search` to use the same API request class. The result of the query is then parsed to provide output to users.
```

Comment:
> Instead of adding a new set of classes and methods, I will modify the methods and classes used by `dotnet tool search` to do a general search when needed. This will allow both `dotnet package search` and `dotnet tool search` to use the same API request class. The result of the query is then parsed to provide output to users.

I don't think this is the right approach, if you mean what I think this is saying. The `dotnet tool search` code was added such a long time ago, I don't remember if they asked us, the NuGet Client team, for an API change to support the new query syntax that nuget.org added. Maybe they didn't ask, because what we add into NuGet.Client is generalised to support any NuGet server implementation, not just nuget.org, and this package type filtering is specific to nuget.org.  Maybe the SDK/CLI team did ask us, but we didn't commit to doing it quickly enough for them, so they decided to re-implement parts of NuGet.Protocol themselves to unblock their scenario.

Anyway, I feel like we should be encouraging re-use of our NuGet.Protocol package, not custom implementations of the NuGet server API.

Our docs on the NuGet Client SDK have an almost complete example of how `dotnet package search` should be implemented: https://learn.microsoft.com/en-us/nuget/reference/nuget-client-sdk#search-packages. It's probably more or less the same code that `nuget.exe search` has. No need to extend dotnet/sdk's existing code to support what our code already supports.

It would be great for us, the NuGet.Client team, to add relevant APIs into NuGet.Protocol's search to do package type filtering (it will be needed if I get my way with https://github.com/NuGet/Home/issues/12567), at which time the dotnet/sdk repo can remove their custom NuGet protocol API code. However, adding package type filtering into NuGet.Protocol is out of scope of this spec on adding `dotnet package search`.

Reply:
Thanks Andy, I have changed my implementation to use Nuget.Protocol and left dotnet tool search as it is.